### ISSUE TITLE
Is the overhead introduced by the deep learning frameworks significant or not?

### ISSUE THREAD

**Issue by snarb** [01org/mkl-dnn#105]
Hi guys! Is there a difference in perfomance when using mkl- dnn directly vs some supported deep learning framework for example TensorFlow or Caffe on the last Intel CPUs? For example for AlexNet. How significant is overhead introduced by the deep lerning framework layer? Thanks.

**Comment by emfomenk** [comment]
Hi Pavel,

Different frameworks may have different overheads. Mainly they come from:

- 'completeness' of the integration: MKL-DNN requires special attention with data conversions. As you know to obtain good performance one should avoid as many unnecessary conversions as possible. But implementing that might be tricky. So depending on how good the integration the performance for the same topology might be different. AFAIK IntelCaffe has pretty robust/decent integration and almost the avoidable conversions would be avoided.

- FW peculiarities. Sometimes the FW's layer may not be directly expressed with MKL-DNN primitive. In that case some overhead appears for data preparation or something similar. An example is scale'n'shift parameter in batch-normalization layer in IntelCaffe. In the FW those to parameters are kept in separate fashion, while MKL-DNN requires those to lie one after another. Hence some overhead [appears](https://github.com/intel/caffe/blob/master/src/caffe/layers/mkldnn_batch_norm_layer.cpp#L305).

- Finally, not all the primitives are available in MKL-DNN, so FW relies on its own implementation. Strictly speaking this not an overhead, but something that should be kept in mind.

Long story short, IntelCaffe has quite small overhead. For Alexnet topology if I am not mistaken IntelCaffe spends around 95% of the time in MKL-DNN, which allows the framework leverage MKL-DNN optimizations in a very optimal way.

**Comment by qgao007** [comment]
Hi emfomenk, when I use benchdnn vs framework (verbose) for the same shapes, will the overhead you mentioned above be included in the framework (verbose) time duration? I assume it is, right?

**Comment by dzarukin** [comment]
Hi @qgao007, thank you for the question.

Verbose time measures only oneDNN performance, the overhead time is not included. But numbers reported by verbose may vary because of different memory access patterns.

In benchdnn performance validation targeting kernel efficiency estimation in best conditions when memory is looped and re-used which can lead to best way of utilizing caches. Such effect may not happen in a real model on framework side and numbers will be less attractive. Thank you.

**Comment by qgao007** [comment]
Hi @dzarukin, thanks for sharing your insights. Very helpful for my benchmarking work. Some follow-up questions. 
1. The memory access patterns in cache is only applied to kernel and not related to data input/output, right? 
2. What do you expect the impact of multi-threading on FW overhead will be? If multi-threading does matters how do you expect OpenMP vs. Eigen in the sense of overhead? I guess OpenMP should be better?

**Comment by dzarukin** [comment]
It is directly related to data input and output.
Benchdnn re-uses same memories for running the execution part. It means that if the total amount fits the cache, data will be in cache.
Real models might have data cached if only the previous operation used oneDNN and the amount of destination memory was able to fit cache for the next primitive.

The term "overhead" is very broad and can't be estimated as is. May I ask why are you really interested in "overhead"?

**Comment by qgao007** [comment]
Thanks @dzarukin. Happy to answer your question. I want to compare performance between different frameworks that both use oneDNN. So for the same oneDNN (same shapes) verbose, if there is a duration time gap between frameworks, it's of big interest to me to drill down and find out what's causing that, such as primitive cache implementation (you highlighted) or something else (eg. multithreading implementation)? Please let me know if anything is unclear.
